homeassistant/components/gtfs/sensor.py

```diff
@@ -308,123 +382,303 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+
+            # Fetch next departure
             self._departure = get_next_departure(
                 self._pygtfs, self.origin, self.destination, self._offset,
                 self._include_tomorrow)
+
+            # Define the state as a UTC timestamp with ISO 8601 format
             if not self._departure:
                 self._state = None
-                self._attributes = {}
-                self._attributes['Info'] = "No more departures" if \
-                    self._include_tomorrow else "No more departures today"
-                if self._name == '':
-                    self._name = (self._custom_name or DEFAULT_NAME)
-                return
-
-            # Define the state as a UTC timestamp with ISO 8601 format.
-            arrival_time = dt_util.as_utc(
-                self._departure['arrival_time']).isoformat()
-            departure_time = dt_util.as_utc(
-                self._departure['departure_time']).isoformat()
-            self._state = departure_time
-
-            origin_station = self._departure['origin_station']
-            destination_station = self._departure['destination_station']
-            origin_stop_time = self._departure['origin_stop_time']
-            destination_stop_time = self._departure['destination_stop_time']
-            agency = self._departure['agency']
-            route = self._departure['route']
-            trip = self._departure['trip']
-
-            name = '{} {} to {} next departure'
+            else:
+                self._state = dt_util.as_utc(
+                    self._departure['departure_time']).isoformat()
+
+            # Fetch trip and route details once, unless updated
+            if not self._departure:
+                self._trip = None
+            else:
+                trip_id = self._departure['trip_id']
+                if not self._trip or self._trip.trip_id != trip_id:
+                    _LOGGER.info("Fetching trip details for %s", trip_id)
+                    self._trip = self._pygtfs.trips_by_id(trip_id)[0]
+
+                route_id = self._departure['route_id']
+                if not self._route or self._route.route_id != route_id:
+                    _LOGGER.info("Fetching route details for %s", route_id)
+                    self._route = self._pygtfs.routes_by_id(route_id)[0]
+
+            # Fetch agency details exactly once
+            if self._agency is None and self._route:
+                try:
+                    _LOGGER.info("Fetching agency details for %s",
+                                 self._route.agency_id)
+                    self._agency = self._pygtfs.agencies_by_id(
+                        self._route.agency_id)[0]
+                except IndexError:
+                    _LOGGER.warning(
+                        "Agency ID '%s' not found in agency table. You may "
+                        "want to update the agency database table to fix this "
+                        "missing reference.", self._route.agency_id)
```

Comment:
We don't end logging messages with period.

Reply:
Will fix in next PR!